File: packages/ce-core/src/cache.ts
import type { TokenEstimator } from "./types.js";

interface CacheOptions {
  maxSize?: number;
}

/**
 * Create a cached token estimator using an LRU cache.
 *
 * Wraps an existing estimator with content-keyed caching to avoid
 * redundant estimation of the same text.
 *
 * @param estimator - The base token estimator to wrap
 * @param options - Cache options (maxSize defaults to 1000)
 * @returns A cached TokenEstimator
 *
 * @example
 * ```ts
 * import { createCachedEstimator } from "@ce/core";
 * import { openaiTokenEstimator } from "@ce/providers";
 *
 * const cached = createCachedEstimator(openaiTokenEstimator, { maxSize: 500 });
 * pack(items, budget, { tokenEstimator: cached });
 * ```
 */
export function createCachedEstimator(
  estimator: TokenEstimator,
  options: CacheOptions = {}
): TokenEstimator {
  const maxSize = options.maxSize ?? 1000;
  const cache = new Map<string, number>();

  return (text: string, opts?: { model?: string; provider?: string }) => {
    const key =
      opts?.model || opts?.provider
        ? `${text}\0${opts.model ?? ""}\0${opts.provider ?? ""}`
        : text;
    if (cache.has(key)) {
      return cache.get(key)!;
    }

    const result = estimator(text, opts);

    if (cache.size >= maxSize) {
      const firstKey = cache.keys().next().value;
      if (firstKey !== undefined) {
        cache.delete(firstKey);
      }
    }

    cache.set(key, result);
    return result;
  };
}

File: python/context_engineering/cache.py
"""Cache module: LRU-cached token estimator wrapper.

Wraps an existing estimator with content-keyed caching to avoid
redundant estimation of the same text.
"""

from __future__ import annotations

import threading
from collections import OrderedDict
from typing import Callable


def create_cached_estimator(
    estimator: Callable[[str], int],
    max_size: int = 1000,
) -> Callable[[str], int]:
    """Create a cached token estimator using an LRU cache.

    Args:
        estimator: The base token estimator to wrap.
        max_size: Maximum cache entries (default: 1000).

    Returns:
        A cached estimator function with the same signature.
    """
    cache: OrderedDict[str, int] = OrderedDict()
    lock = threading.Lock()

    def cached(text: str) -> int:
        with lock:
            if text in cache:
                cache.move_to_end(text)
                return cache[text]

        result = estimator(text)

        with lock:
            if len(cache) >= max_size:
                cache.popitem(last=False)

            cache[text] = result
        return result

    return cached
